DDF-4031 As a user I would like to be able to export a result set as a KML file so that I can have a kml file to use externally

Corey-Collins opened the pull request:
<!--
Uncomment this block and include PR # of change against 2.13.x where appropriate.
When doing so, it is reasonable to delete much of the rest of the PR description, leaving only:
- This uncommented block with link to 2.13.x PR
- Reviewers tagged
- Teams tagged
- Committers tagged
- The link to the "Notes on Review Process"

##### ABBREVIATED REVIEW BETWEEN 2.13.X AND MASTER IS IN EFFECT
Link to 2.13.x PR: #XXXX
____
-->

#### What does this PR do?
The current KMLTransformerImpl (QueryResponseTransformer) exports attributes to the description tag. This also extracts attributes and injects them into extended data. 
#### Who is reviewing it? 
<!--(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)-->
@Schachte 
@brianfelix 
@Bdthomson 
@jrnorth 
@bellcc
@willwill96 
#### Select relevant component teams: 
https://github.com/orgs/codice/teams
#### Ask 2 committers to review/merge the PR and tag them here.
<!--
If you don't know who to ask, you can request reviews in https://groups.google.com/forum/#!forum/ddf-developers .
(please choose ONLY two committers from below, delete the rest)
-->
@millerw8 
@bdeining
#### How should this be tested?
<!--(List steps with links to updated documentation)-->
- Perform `/internal/cql/transform/kml` query with [DDF-3997](https://github.com/codice/ddf/pull/3516) and ensure KML file is in attachment header (web sockets off, populated with metacards).
- Validate KML data, including description and extended data.
#### Any background context you want to provide?
Relies on [DDF-3997](https://github.com/codice/ddf/pull/3516)
#### What are the relevant tickets?
[](https://codice.atlassian.net/browse/)
#### Screenshots
<!--(if appropriate)-->
#### Checklist:
- [ ] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

willwill96 commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +221,91 @@ protected Placemark performDefaultTransformation(Metacard entry)
+    return data;
+  }
+
+  private static Object convertAttribute(Attribute attribute, AttributeDescriptor descriptor) {
```
nit - I think it might be better to return a list of serializables and return a list of size 1 in the case of a single value, just to make method declarations more clear.

bellcc replied:
✏️ This method should only be static if it makes sense to call this method before an object has been instantiated and since it's a private method that's probably not the case

willwill96 commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +221,91 @@ protected Placemark performDefaultTransformation(Metacard entry)
+    }
+  }
+
+  private static Object convertValue(
```
nit - I think returning a serializable is more appropriate.

bellcc commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +221,91 @@ protected Placemark performDefaultTransformation(Metacard entry)
+        // class variable. If this proves to be a slowdown this class should be refactored
+        // such that we don't need this method to be static.
+        SimpleDateFormat dateFormat = new SimpleDateFormat(ISO_8601_DATE_FORMAT);
+        dateFormat.setTimeZone(TimeZone.getTimeZone("GMT"));
```
❓ I know that UTC and GMT but should we stick to UTC for consistency? 

`TimeZone.getTimeZone("UTC")`

bellcc commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +221,92 @@ protected Placemark performDefaultTransformation(Metacard entry)
+    }
+  }
+
+  private static Serializable convertValue(
```
✏️ Same thing as above

bellcc commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +221,92 @@ protected Placemark performDefaultTransformation(Metacard entry)
+    return data;
+  }
+
+  private static Serializable convertAttribute(
```
✏️ This method should only be static if it makes sense to call this method before an object has been instantiated and since it's a private method that's probably not the case

bellcc commented on catalog/spatial/kml/spatial-kml-transformer/src/main/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -17,14 +17,14 @@
-		
-		<AD name="Style URL" id="styleUrl" required="true" type="String"
-            description="The full qualified URL to the KML Style. e.g. http://example.com/styles#myStyle"/>
+		<AD name="Attribute Name" id="attributeName" type="String"
```
❓ `True` is the default value for these attribute definitions so this change doesn't do anything but I still like to have it to make it explicit

jrnorth commented:
So multivalued attributes will have comma-separated strings in their value elements? Does KML not support multiple values per data element?

jrnorth commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +221,92 @@ protected Placemark performDefaultTransformation(Metacard entry)
+        // it is not a threadsafe class so we are not able to put it in a static
+        // class variable. If this proves to be a slowdown this class should be refactored
+        // such that we don't need this method to be static.
+        SimpleDateFormat dateFormat = new SimpleDateFormat(ISO_8601_DATE_FORMAT);
```
You can use [`DateTimeFormatter`](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html) instead, which is thread-safe. To format a `Date` you would just have to call `formatter.format(((Date) value).toInstant())` instead.

Corey-Collins commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +221,91 @@ protected Placemark performDefaultTransformation(Metacard entry)
+            convertAttribute(attribute, attributeDescriptor)
+                .toString()
+                .replace("[", "")
+                .replace("]", "");
```
@willwill96 i think using `.substring(1, str.length()-1)` would simplify all of this

Corey-Collins replied:
@willwill96 that actually won't work since brackets are only added in the case the attribute is a list

Corey-Collins replied:
@willwill96 
`StringUtils.removeStart(StringUtils.removeEnd(string, "]"), "[");`
seems to work best and is null safe

Corey-Collins commented:
@jrnorth 

> So multivalued attributes will have comma-separated strings in their value elements? Does KML not support multiple values per data element?

That's what I've been researching with this ticket, and this has seemed to be the best solution so far given how the library inputs `Data` objects into extended data, which are essentially just key/value pairs.

Also to note that in the future the exact format of KML may need to be changed in a later ticket, the main purpose in this ticket is to support exporting with the `cql/transform/{id}` endpoint.

willwill96 commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +228,85 @@ protected Placemark performDefaultTransformation(Metacard entry)
+      case XML:
+      case FLOAT:
+      case GEOMETRY:
+        return value.toString();
```
nit - Just return value. toString()/valueOf() will be called on it regardless.

bdeining commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +228,85 @@ protected Placemark performDefaultTransformation(Metacard entry)
+      final String attributeName = attributeDescriptor.getName();
+      final Attribute attribute = metacard.getAttribute(attributeName);
+      if (attribute != null) {
+        String attributeValue = convertAttribute(attribute, attributeDescriptor).toString();
```
We should add some null checking here, since `convertAttribute` can return null

bdeining commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +228,85 @@ protected Placemark performDefaultTransformation(Metacard entry)
+  private Serializable convertAttribute(Attribute attribute, AttributeDescriptor descriptor) {
+
+    if (descriptor.isMultiValued()) {
+      ArrayList<Serializable> values = new ArrayList<>();
```
we should declare this as the interface `List`

bdeining commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +228,85 @@ protected Placemark performDefaultTransformation(Metacard entry)
+
+    switch (format) {
+      case DATE:
+        if (!(value instanceof Date)) {
```
Is it possible for the value to not be a date if the attribute is a date type?  I think this would cause a Solr exception on ingest

Corey-Collins replied:
from what i've asked around it doesn't seem possible. very similar code is also being used at https://github.com/codice/ddf/blob/f008f640967fa6a655943f0a8d7068d0c5d063db/catalog/transformer/catalog-transformer-propertyjson-metacard/src/main/java/ddf/catalog/transformer/metacard/propertyjson/PropertyJsonMetacardTransformer.java#L187
hearing more input from others would be helpful on this though

bdeining commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +228,90 @@ protected Placemark performDefaultTransformation(Metacard entry)
+      final String attributeName = attributeDescriptor.getName();
+      final Attribute attribute = metacard.getAttribute(attributeName);
+      if (attribute != null) {
+        Serializable convertAttribute = convertAttribute(attribute, attributeDescriptor);
```
✏️ convertedAttribute might be more appropriate

bdeining commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +228,90 @@ protected Placemark performDefaultTransformation(Metacard entry)
+      case LONG:
+      case INTEGER:
+      case SHORT:
+        return value;
```
can remove this; since line 308 does the same thing

jrnorth commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +228,89 @@ protected Placemark performDefaultTransformation(Metacard entry)
+          LOGGER.debug("Attribute {} converted to null value.", attributeName);
+        } else {
+          String attributeValue = convertedAttribute.toString();
+          attributeValue = StringUtils.removeStart(StringUtils.removeEnd(attributeValue, "]"), "[");
```
It's possible for an attribute (e.g. one with a String value) to begin or end with these characters. Since `convertAttribute` can return either a list or a single value, these could be mistakenly removed. If you performed the String conversion in `convertAttribute` you could strip these out only when the attribute is multivalued.

bdeining commented on catalog/spatial/kml/spatial-kml-transformer/src/test/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImplTest.java:
```diff
@@ -222,6 +248,71 @@ public void testPerformDefaultTransformationMultiLineStringLocation()
+
+    for (Data data : dataList) {
+      switch (data.getName()) {
+        case "id":
```
These attributes have constants in `Core` and `Metacard` that we should use

bdeining commented on catalog/spatial/kml/spatial-kml-transformer/src/test/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImplTest.java:
```diff
@@ -222,6 +248,71 @@ public void testPerformDefaultTransformationMultiLineStringLocation()
+        case "metacard-tags":
+          assertThat(
+              data.getValue(),
+              is(METACARD_TAGS.asList().get(0) + ", " + METACARD_TAGS.asList().get(1)));
```
I think if you made `METACARD_TAGS` a list then you can just `assertThat(data.getValue(), is(METACARD_TAGS)`

bdeining commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +228,94 @@ protected Placemark performDefaultTransformation(Metacard entry)
+        }
+        stringList += values.get(i);
+      }
+      return stringList;
```
I think you can `return StringUtils.join(values, ",");` and cut out this stringlist entirely

bdeining commented on catalog/spatial/kml/spatial-kml-transformer/src/main/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImpl.java:
```diff
@@ -208,6 +228,94 @@ protected Placemark performDefaultTransformation(Metacard entry)
+      List<Serializable> values = new ArrayList<>();
+      for (Serializable value : attribute.getValues()) {
+        values.add(
+            convertValue(attribute.getName(), value, descriptor.getType().getAttributeFormat()));
```
since `convertValue` can return null, `values` can potentially contain null values.

Corey-Collins commented on catalog/spatial/kml/spatial-kml-transformer/src/test/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImplTest.java:
```diff
@@ -222,6 +249,71 @@ public void testPerformDefaultTransformationMultiLineStringLocation()
+        case Core.METACARD_TAGS:
+          assertThat(
+              data.getValue(),
+              is(METACARD_TAGS.asList().get(0) + "," + METACARD_TAGS.asList().get(1)));
```
@bdeining there's some trouble with using `METACARD_TAGS` as a List. `metacard.setTags` requires a Set, and just having `is(METACARD_TAGS)` would return `[item1,item2]` where the actual is `item1,item2`

bdeining commented:
build now

Corey-Collins commented:
build now

Corey-Collins commented:
@millerw8 here's a sample:
```<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<kml:kml xmlns:gx="http://www.google.com/kml/ext/2.2" xmlns:xal="urn:oasis:names:tc:ciq:xsdschema:xAL:2.0" xmlns:kml="http://www.opengis.net/kml/2.2" xmlns:atom="http://www.w3.org/2005/Atom">
    <kml:Document id="2ee2ff21-426f-41eb-9538-1a5293494cdf">
        <kml:name>Results (2)</kml:name>
        <kml:open>0</kml:open>
        <kml:Style id="bluenormal">
            <kml:LabelStyle>
                <kml:scale>0.0</kml:scale>
            </kml:LabelStyle>
            <kml:LineStyle>
                <kml:color>33ff0000</kml:color>
                <kml:width>3.0</kml:width>
            </kml:LineStyle>
            <kml:PolyStyle>
                <kml:color>33ff0000</kml:color>
                <kml:fill>1</kml:fill>
            </kml:PolyStyle>
            <kml:BalloonStyle>
                <kml:text>&lt;h3&gt;&lt;b&gt;$[name]&lt;/b&gt;&lt;/h3&gt;&lt;table&gt;&lt;tr&gt;&lt;td width="400"&gt;$[description]&lt;/td&gt;&lt;/tr&gt;&lt;/table&gt;</kml:text>
            </kml:BalloonStyle>
        </kml:Style>
        <kml:Style id="bluehighlight">
            <kml:LabelStyle>
                <kml:scale>1.0</kml:scale>
            </kml:LabelStyle>
            <kml:LineStyle>
                <kml:color>99ff0000</kml:color>
                <kml:width>6.0</kml:width>
            </kml:LineStyle>
            <kml:PolyStyle>
                <kml:color>99ff0000</kml:color>
                <kml:fill>1</kml:fill>
            </kml:PolyStyle>
            <kml:BalloonStyle>
                <kml:text>&lt;h3&gt;&lt;b&gt;$[name]&lt;/b&gt;&lt;/h3&gt;&lt;table&gt;&lt;tr&gt;&lt;td width="400"&gt;$[description]&lt;/td&gt;&lt;/tr&gt;&lt;/table&gt;</kml:text>
            </kml:BalloonStyle>
        </kml:Style>
        <kml:StyleMap id="default">
            <kml:Pair>
                <kml:key>normal</kml:key>
                <kml:styleUrl>#bluenormal</kml:styleUrl>
            </kml:Pair>
            <kml:Pair>
                <kml:key>highlight</kml:key>
                <kml:styleUrl>#bluehighlight</kml:styleUrl>
            </kml:Pair>
        </kml:StyleMap>
        <kml:Placemark id="Placemark-daf8edfe4c204a088e4dd0ce271cdebc">
            <kml:name>9466484_b06f26d579_o.jpg</kml:name>
            <kml:description>&lt;!DOCTYPE html&gt;
&lt;html&gt;
  &lt;head&gt;
    &lt;meta content="text/html; charset=windows-1252" http-equiv="content-type"&gt;
    &lt;style media="screen" type="text/css"&gt;
      .label {
        font-weight: bold
      }
      .linkTable {
        width: 100%
      }
      .thumbnailDiv {
        text-align: center
      }
      img {
        max-width: 100px;
        max-height: 100px;
        border-style:none
      }
      tr:nth-child(even) {
        background-color: #c0c0c0;
      }
    &lt;/style&gt;
  &lt;/head&gt;
  &lt;body&gt;

    

    &lt;table style="width:100%"&gt;
        &lt;tbody&gt;
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Created&lt;/td&gt;
                &lt;td&gt;2005-04-15T03:35:50-0600&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Effective&lt;/td&gt;
                &lt;td&gt;2018-08-29T15:40:47-0600&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Id&lt;/td&gt;
                &lt;td&gt;daf8edfe4c204a088e4dd0ce271cdebc&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Metacard Tags&lt;/td&gt;
                &lt;td&gt;resource&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Metacard Created&lt;/td&gt;
                &lt;td&gt;2018-08-29T15:40:47-0600&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Metacard Modified&lt;/td&gt;
                &lt;td&gt;2018-08-29T15:40:47-0600&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Modified&lt;/td&gt;
                &lt;td&gt;2005-04-15T05:44:18-0600&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Point Of Contact&lt;/td&gt;
                &lt;td&gt;admin@localhost.local&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Resource Download Url&lt;/td&gt;
                &lt;td&gt;https://localhost:8993/services/catalog/sources/ddf.distribution/daf8edfe4c204a088e4dd0ce271cdebc?transform=resource&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Title&lt;/td&gt;
                &lt;td&gt;9466484_b06f26d579_o.jpg&lt;/td&gt;
            &lt;/tr&gt;
            
        
        &lt;/tbody&gt;
    &lt;/table&gt;
    &lt;table class="linkTable"&gt;
      &lt;tr&gt;
        &lt;td&gt;&lt;a href="https://localhost:8993/services/catalog/sources/ddf.distribution/daf8edfe4c204a088e4dd0ce271cdebc?transform=resource"&gt;Download...&lt;/a&gt;&lt;/td&gt;
          &lt;td&gt;&lt;/td&gt;
      &lt;/tr&gt;
    &lt;/table&gt;
  &lt;/body&gt;
&lt;/html&gt;
</kml:description>
            <kml:TimeSpan>
                <kml:begin>2018-08-29T21:40:47</kml:begin>
            </kml:TimeSpan>
            <kml:styleUrl>#default</kml:styleUrl>
            <kml:ExtendedData>
                <kml:Data name="metacard-tags">
                    <kml:value>resource,VALID</kml:value>
                </kml:Data>
                <kml:Data name="resource-download-url">
                    <kml:value>https://localhost:8993/services/catalog/sources/ddf.distribution/daf8edfe4c204a088e4dd0ce271cdebc?transform=resource</kml:value>
                </kml:Data>
                <kml:Data name="modified">
                    <kml:value>2005-04-15T11:44:18.000+0000</kml:value>
                </kml:Data>
                <kml:Data name="location">
                    <kml:value>POINT (10.402439 63.418399)</kml:value>
                </kml:Data>
                <kml:Data name="title">
                    <kml:value>9466484_b06f26d579_o.jpg</kml:value>
                </kml:Data>
                <kml:Data name="point-of-contact">
                    <kml:value>admin@localhost.local</kml:value>
                </kml:Data>
                <kml:Data name="metacard.modified">
                    <kml:value>2018-08-29T21:40:47.590+0000</kml:value>
                </kml:Data>
                <kml:Data name="created">
                    <kml:value>2005-04-15T09:35:50.000+0000</kml:value>
                </kml:Data>
                <kml:Data name="effective">
                    <kml:value>2018-08-29T21:40:47.590+0000</kml:value>
                </kml:Data>
                <kml:Data name="metacard.created">
                    <kml:value>2018-08-29T21:40:47.590+0000</kml:value>
                </kml:Data>
                <kml:Data name="resource-uri">
                    <kml:value>content:daf8edfe4c204a088e4dd0ce271cdebc</kml:value>
                </kml:Data>
                <kml:Data name="id">
                    <kml:value>daf8edfe4c204a088e4dd0ce271cdebc</kml:value>
                </kml:Data>
            </kml:ExtendedData>
            <kml:Point>
                <kml:coordinates>10.402439,63.418399</kml:coordinates>
            </kml:Point>
        </kml:Placemark>
        <kml:Placemark id="Placemark-daf8edfe4c204a088e4dd0ce271cdebd">
            <kml:name>9466484_b06f26d579_o.jpg</kml:name>
            <kml:description>&lt;!DOCTYPE html&gt;
&lt;html&gt;
  &lt;head&gt;
    &lt;meta content="text/html; charset=windows-1252" http-equiv="content-type"&gt;
    &lt;style media="screen" type="text/css"&gt;
      .label {
        font-weight: bold
      }
      .linkTable {
        width: 100%
      }
      .thumbnailDiv {
        text-align: center
      }
      img {
        max-width: 100px;
        max-height: 100px;
        border-style:none
      }
      tr:nth-child(even) {
        background-color: #c0c0c0;
      }
    &lt;/style&gt;
  &lt;/head&gt;
  &lt;body&gt;

    

    &lt;table style="width:100%"&gt;
        &lt;tbody&gt;
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Created&lt;/td&gt;
                &lt;td&gt;2005-04-15T03:35:50-0600&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Description&lt;/td&gt;
                &lt;td&gt;red house of brick&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Effective&lt;/td&gt;
                &lt;td&gt;2018-08-29T15:40:47-0600&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Id&lt;/td&gt;
                &lt;td&gt;daf8edfe4c204a088e4dd0ce271cdebd&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Metacard Tags&lt;/td&gt;
                &lt;td&gt;resource&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Metacard Created&lt;/td&gt;
                &lt;td&gt;2018-08-29T15:40:47-0600&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Metacard Modified&lt;/td&gt;
                &lt;td&gt;2018-08-29T15:40:47-0600&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Modified&lt;/td&gt;
                &lt;td&gt;2005-04-15T05:44:18-0600&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Point Of Contact&lt;/td&gt;
                &lt;td&gt;admin@localhost.local&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Resource Download Url&lt;/td&gt;
                &lt;td&gt;https://localhost:8993/services/catalog/sources/ddf.distribution/daf8edfe4c204a088e4dd0ce271cdebd?transform=resource&lt;/td&gt;
            &lt;/tr&gt;
            
        
            
        
            
            &lt;tr&gt;
                &lt;td style="font-weight: bold;"&gt;Title&lt;/td&gt;
                &lt;td&gt;9466484_b06f26d579_o.jpg&lt;/td&gt;
            &lt;/tr&gt;
            
        
        &lt;/tbody&gt;
    &lt;/table&gt;
    &lt;table class="linkTable"&gt;
      &lt;tr&gt;
        &lt;td&gt;&lt;a href="https://localhost:8993/services/catalog/sources/ddf.distribution/daf8edfe4c204a088e4dd0ce271cdebd?transform=resource"&gt;Download...&lt;/a&gt;&lt;/td&gt;
          &lt;td&gt;&lt;/td&gt;
      &lt;/tr&gt;
    &lt;/table&gt;
  &lt;/body&gt;
&lt;/html&gt;
</kml:description>
            <kml:TimeSpan>
                <kml:begin>2018-08-29T21:40:47</kml:begin>
            </kml:TimeSpan>
            <kml:styleUrl>#default</kml:styleUrl>
            <kml:ExtendedData>
                <kml:Data name="metacard-tags">
                    <kml:value>resource,VALID</kml:value>
                </kml:Data>
                <kml:Data name="description">
                    <kml:value>red house of brick</kml:value>
                </kml:Data>
                <kml:Data name="resource-download-url">
                    <kml:value>https://localhost:8993/services/catalog/sources/ddf.distribution/daf8edfe4c204a088e4dd0ce271cdebd?transform=resource</kml:value>
                </kml:Data>
                <kml:Data name="modified">
                    <kml:value>2005-04-15T11:44:18.000+0000</kml:value>
                </kml:Data>
                <kml:Data name="location">
                    <kml:value>POINT (10.402439 63.418399)</kml:value>
                </kml:Data>
                <kml:Data name="title">
                    <kml:value>9466484_b06f26d579_o.jpg</kml:value>
                </kml:Data>
                <kml:Data name="point-of-contact">
                    <kml:value>admin@localhost.local</kml:value>
                </kml:Data>
                <kml:Data name="metacard.modified">
                    <kml:value>2018-08-29T21:40:47.285+0000</kml:value>
                </kml:Data>
                <kml:Data name="created">
                    <kml:value>2005-04-15T09:35:50.000+0000</kml:value>
                </kml:Data>
                <kml:Data name="effective">
                    <kml:value>2018-08-29T21:40:47.285+0000</kml:value>
                </kml:Data>
                <kml:Data name="metacard.created">
                    <kml:value>2018-08-29T21:40:47.285+0000</kml:value>
                </kml:Data>
                <kml:Data name="resource-uri">
                    <kml:value>content:daf8edfe4c204a088e4dd0ce271cdebc</kml:value>
                </kml:Data>
                <kml:Data name="id">
                    <kml:value>daf8edfe4c204a088e4dd0ce271cdebd</kml:value>
                </kml:Data>
            </kml:ExtendedData>
            <kml:Point>
                <kml:coordinates>10.402439,63.418399</kml:coordinates>
            </kml:Point>
        </kml:Placemark>
    </kml:Document>
</kml:kml>```

jrnorth commented on catalog/spatial/kml/spatial-kml-transformer/src/test/java/org/codice/ddf/spatial/kml/transformer/KMLTransformerImplTest.java:
```diff
@@ -222,6 +254,71 @@ public void testPerformDefaultTransformationMultiLineStringLocation()
+        case Core.EXPIRATION:
+        case Core.CREATED:
+          assertThat(data.getValue(), is(METACARD_DATE_STRING));
+          break;
```
✏️ Since there's no default case, this test will pass even if there are unexpected entries in the extended data. Consider adding a default case that fails the test.

jrnorth replied:
Fixing this will also resolve your Codacy issue.

Corey-Collins commented:
build now

hayleynorton commented:
Hero (quick) build successful
![image](https://user-images.githubusercontent.com/39737329/44866584-e71c2300-ac42-11e8-99d2-a343dc3b2e50.png)
![image](https://user-images.githubusercontent.com/39737329/44866607-f4d1a880-ac42-11e8-86b2-aa57137eae71.png)
![image](https://user-images.githubusercontent.com/39737329/44866725-44b06f80-ac43-11e8-9d9c-1b3eab5938c5.png)
